File: android/app/src/main/java/com/sujayarvind/dax90/DaXWidget.java
package com.sujayarvind.dax90;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.widget.RemoteViews;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DaXWidget extends AppWidgetProvider {
    public static final String PREFS = "dax";
    public static final String STATE_KEY = "state";

    @Override
    public void onUpdate(Context context, AppWidgetManager manager, int[] appWidgetIds) {
        for (int id : appWidgetIds) updateWidget(context, manager, id);
    }

    public static void updateAll(Context context) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        ComponentName name = new ComponentName(context, DaXWidget.class);
        int[] ids = manager.getAppWidgetIds(name);
        for (int id : ids) updateWidget(context, manager, id);
    }

    private static void updateWidget(Context context, AppWidgetManager manager, int id) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
        String raw = prefs.getString(STATE_KEY, "");
        String done = match(raw, "\\\"doneCount\\\"\\s*:\\s*(\\d+)", "0");
        String total = match(raw, "\\\"total\\\"\\s*:\\s*(\\d+)", "13");
        String current = match(raw, "\\\"current\\\"\\s*:\\s*\\\"([^\\\"]*)\\\"", "Open 90DaX");
        int percent = 0;
        try {
            int d = Integer.parseInt(done), t = Math.max(1, Integer.parseInt(total));
            percent = Math.round(d * 100f / t);
        } catch (Exception ignored) { }

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.dax_widget);
        views.setTextViewText(R.id.widget_title, "90DaX");
        views.setTextViewText(R.id.widget_progress, done + " / " + total + " activities");
        views.setTextViewText(R.id.widget_current, current);
        views.setTextViewText(R.id.widget_percent, percent + "% complete");
        Intent launch = new Intent(context, MainActivity.class);
        views.setOnClickPendingIntent(R.id.widget_root,
                android.app.PendingIntent.getActivity(context, 0, launch,
                        android.app.PendingIntent.FLAG_UPDATE_CURRENT | android.app.PendingIntent.FLAG_IMMUTABLE));
        manager.updateAppWidget(id, views);
    }

    private static String match(String input, String regex, String fallback) {
        Matcher m = Pattern.compile(regex).matcher(input == null ? "" : input);
        return m.find() ? m.group(1) : fallback;
    }
}

File: android/app/src/main/java/com/sujayarvind/dax90/DaXWidget.kt
package com.sujayarvind.dax90

import android.content.Context
import android.graphics.Color
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.action.actionStartActivity
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.appwidget.updateAll
import androidx.glance.background
import androidx.glance.layout.Alignment
import androidx.glance.layout.Column
import androidx.glance.layout.Row
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.padding
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import androidx.glance.unit.ColorProvider
import androidx.glance.unit.dp
import androidx.glance.unit.sp

class DaXWidget : GlanceAppWidget() {
    override suspend fun provideGlance(context: Context, id: GlanceId) {
        val raw = context.getSharedPreferences("dax", 0).getString("state", null) ?: ""
        val done = Regex("\\\"doneCount\\\"\\s*:\\s*(\\d+)").find(raw)?.groupValues?.getOrNull(1) ?: "0"
        val total = Regex("\\\"total\\\"\\s*:\\s*(\\d+)").find(raw)?.groupValues?.getOrNull(1) ?: "13"
        val current = Regex("\\\"current\\\"\\s*:\\s*\\\"([^\\\"]*)\\\"").find(raw)?.groupValues?.getOrNull(1) ?: "Open 90DaX"
        provideContent {
            Column(modifier = GlanceModifier.fillMaxSize().background(ColorProvider(Color.rgb(9,11,10))).padding(14.dp), verticalAlignment = Alignment.CenterVertically) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text("90DaX", style = TextStyle(color = ColorProvider(Color.WHITE), fontSize = 18.sp))
                    Text("  •  TODAY", style = TextStyle(color = ColorProvider(Color.LTGRAY), fontSize = 10.sp), modifier = GlanceModifier.padding(start = 5.dp))
                }
                Text("$done / $total activities", style = TextStyle(color = ColorProvider(Color.WHITE), fontSize = 15.sp), modifier = GlanceModifier.padding(top = 6.dp))
                Text(current, style = TextStyle(color = ColorProvider(Color.LTGRAY), fontSize = 12.sp), modifier = GlanceModifier.padding(top = 4.dp))
                Text("Tap to open", style = TextStyle(color = ColorProvider(Color.WHITE), fontSize = 11.sp), modifier = GlanceModifier.padding(top = 8.dp).clickable(actionStartActivity<MainActivity>()))
            }
        }
    }
}

class DaXWidgetReceiver : GlanceAppWidgetReceiver() {
    override val glanceAppWidget: GlanceAppWidget = DaXWidget()
}
